src/main/java/org/commcare/formplayer/services/MenuSessionRunnerService.java

```diff
@@ -181,6 +190,40 @@ private BaseResponseBean getNextMenu(MenuSession menuSession,
         return menuResponseBean;
     }
 
+    private BaseResponseBean getSmartResponse(MenuSession menuSession, String[] smartLinkParams) {
+        if (smartLinkParams == null || smartLinkParams.length == 0) {
+            return null;
+        }
+
+        String template = storageFactory.getPropertyManager().getSmartLinkTemplate();
+        if (template != null) {
+            SessionWrapper session = menuSession.getSessionWrapper();
+            String command = session.getCommand();
+            if (command != null) {
+                Endpoint endpoint = menuSession.getEndpointByCommand(command);
+                if (endpoint != null) {
+                    for (String param : smartLinkParams) {
+                        template = template.replaceFirst("---", param);
+                    }
+                    template = template.replace("---", endpoint.getId());
```

Comment:
It's a bit weird that an arbitrary number of parameters come from web apps and then one last parameter is set here. I'm debating making each of the param list attributes into just a single param. That would also make it a bit easier to turn that one web-apps-provided param into an app profile property, as described in previous comment.

Reply:
I think you'd want to use a URL template:

```java
UriComponentsBuilder urlBuilder = UriComponentsBuilder.fromHttpUrl(template)
 for (String key : data.keySet()) {
    if (endpoint.getArguments().contains(key)) {
        urlBuilder.queryParam(key, data.get(key));
    }
}
URI url = urlBuilder.build(urlArgs);
```

`urlArgs` could be a list / array or it could be a Map of named params. I think a Map is better but it would require making `smartLinkParams` into a map instead of just a list.

https://docs.spring.io/spring-framework/docs/current/reference/html/web.html#web-uricomponents

Reply:
updated, both to use a template and turn params into a map